app/src/main/java/org/openobservatory/ooniprobe/common/MKCollectorResubmitTask.java

```diff
@@ -0,0 +1,87 @@
+package org.openobservatory.ooniprobe.common;
+
+import android.content.Context;
+import android.util.Log;
+import android.view.WindowManager;
+
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonParser;
+
+import org.openobservatory.ooniprobe.R;
+import org.openobservatory.ooniprobe.dao.MeasurementDao;
+import org.openobservatory.ooniprobe.model.database.Measurement;
+
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.List;
+
+import androidx.appcompat.app.AppCompatActivity;
+import io.ooni.mk.MKCollectorResubmitResults;
+import io.ooni.mk.MKCollectorResubmitSettings;
+import localhost.toolkit.os.NetworkProgressAsyncTask;
+
+public class MKCollectorResubmitTask<A extends AppCompatActivity> extends NetworkProgressAsyncTask<A, Integer, Void> {
+	/**
+	 * {@code new MKCollectorResubmitTask(activity).execute(result_id, measurement_id);}
+	 *
+	 * @param activity from this task are executed
+	 */
+	public MKCollectorResubmitTask(A activity) {
+		super(activity, true, false);
+	}
+
+	private static void perform(Context c, Measurement m) throws IOException {
+		FileInputStream is = new FileInputStream(Measurement.getEntryFile(c, m.id, m.test_name));
+		String input = new GsonBuilder().disableHtmlEscaping().create().toJson(new JsonParser().parse(new InputStreamReader(is)));
+		is.close();
+		MKCollectorResubmitSettings settings = new MKCollectorResubmitSettings();
+		settings.setTimeout(14);
+		settings.setCABundlePath(c.getCacheDir() + "/" + Application.CA_BUNDLE);
+		settings.setSerializedMeasurement(input);
+		MKCollectorResubmitResults results = settings.perform();
+		if (results.isGood()) {
+			String output = results.getUpdatedSerializedMeasurement();
+			FileOutputStream os = new FileOutputStream(Measurement.getEntryFile(c, m.id, m.test_name));
+			os.write(output.getBytes());
+			os.close();
+			m.report_id = results.getUpdatedReportID();
+			m.is_uploaded = true;
+			m.is_upload_failed = false;
+			m.save();
+		} else {
+			Log.w(MKCollectorResubmitSettings.class.getSimpleName(), results.getLogs());
+			// TODO check what to do with logs
+		}
+	}
+
+	@Override protected void onPreExecute() {
+		super.onPreExecute();
+		if (getActivity() != null)
```

Comment:
lifecycle os management start only on idle of the main thread. so in this case there is no difference.

Reply:
I’d rather write robust code that requires less knowledge anyways, especially when doing that is easy and especially when the result requires the reader to know less, not more, about the internals of the systems and about whether they did change.

Reply:
Please, use this pattern. I believe in general it's more OOP, because we try and then see if we could rather than requesting permission beforehand! :-)